### PR TITLE
Add Nesterov Momentum

### PR DESCRIPTION
**Pull request by ajbrock**
This pull request adds the Nesterov Momentum option to SGD, using the parameterization described [here](https://github.com/lisa-lab/pylearn2/pull/136#issuecomment-10381617). I'm not sure if any of the tests need to be modified to include this, but the code here should be correct (tested and working on my CIFAR100 testbed).

### CODE REVIEW

File: torch/optim/sgd.py [hunk 1]
```diff
@@ -51,7 +51,10 @@ def step(self, closure=None):
                         param_state['momentum_buffer'] = d_p.clone()
                     else:
                         buf = param_state['momentum_buffer']
-                        d_p = buf.mul_(momentum).add_(1 - dampening, d_p)
+                        if nesterov !=0:
```
Comment (apaszke) [inline torch/optim/sgd.py]: How about this:
```python
buf.mul_(momentum).add_(1 - dampening, d_p)
if nesterov:
    d_p.add_(momentum, buf)
else:
    d_p = buf
```

File: torch/optim/sgd.py [hunk 2]
```diff
@@ -3,31 +3,30 @@
...
-
         >>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
         >>> optimizer.zero_grad()
         >>> loss_fn(model(input), target).backward()
         >>> optimizer.step()
     """
 
     def __init__(self, params, lr=required, momentum=0, dampening=0,
-                 weight_decay=0):
+                 weight_decay=0, nesterov=False):
```
Comment (apaszke) [inline torch/optim/sgd.py]: Can you please add a check that if `nesterov` is True, then `momentum` has be be greater than 0?

**Review by apaszke — COMMENTED** [review]
There's one more problem - serialized state dicts or optimizers won't have the `nesterov` attribute in any parameter group. I'll figure out how to solve this tomorrow, and I'll get back to you then. Thanks for the PR!

File: torch/optim/sgd.py (hunk 2 above)
Comment (ajbrock) [inline torch/optim/sgd.py]: Should this check also include a condition that dampening be zero, as in the [legacy](https://github.com/ajbrock/pytorch/blob/master/torch/legacy/optim/sgd.py#L44) version?

File: torch/optim/sgd.py (hunk 1 above)
Comment (ajbrock) [inline torch/optim/sgd.py]: Looks good, I'll change it to this.

**Comment by apaszke** [comment]
Pasted from #810:

Ok, so I thought this through and I think the best way would be to implement __setstate__ such that it calls to super and then patches the param groups like that:

```python
def __setstate__(self, state):
    super(RMSprop, self).__setstate__(state)
    for group in self.parameter_groups:
        group.set_default('momentum', 0)
        group.set_default('centered', False)
```

Sorry for waiting!

**Comment by apaszke** [comment]
I added the `__setstate__` and opened a new PR: #887. Thanks!

**Pull request closed without merging**